[DataPipe] Disable profiler for IterDataPipe by default

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #78674

Previously, by default, every call to `iter` and `next` of `IterDataPipe` will have `torch.autograd.profiler.record_function()` as the context. This creates overhead and somewhat slows `IterDataPipe` down.

This PR proposes to disable that by default with a new function to enable it as necessary. I am open to the possibility that perhaps the default should `True` (enabled). Let me know what potential downside this may causes beyond what I understand.

Differential Revision: [D36830074](https://our.internmc.facebook.com/intern/diff/D36830074)

NivekT committed:
[DataPipe] Disable profiler for IterDataPipe by default

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Disable profiler for IterDataPipe by default"




[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36830074).

VitalyFedyunin commented:
I agree to disable it, but in this case, we need something similar to torch enabling mechanisms
```python
with profiler.profile(with_stack=True, profile_memory=True) as prof:
```
Also some sort of documentation on how to profile.

NivekT commented:
> I agree to disable it, but in this case, we need something similar to torch enabling mechanisms
> 
> ```python
> with profiler.profile(with_stack=True, profile_memory=True) as prof:
> ```
> 
> Also some sort of documentation on how to profile.

Based on the current implementation in this PR, it will be:

```python
from torch.utils.data import set_profiler_for_iterdatapipe

# Enable profiling specifically for IterDataPipe
torch.utils.data.set_profiler_for_iterdatapipe(enable=True)

with torch.profiler.profile(...) as p:
    # Rest of code
```

With the latter part look something like this:
https://github.com/pytorch/kineto/blob/plugin/vnext/tb_plugin/examples/datapipe_example.py


Agree on the need for documentation. Does this seem reasonable to you?

ejguan commented:
Is that possible for TorchData to re-use the context manager of 

> ```python
> with profiler.profile(with_stack=True, profile_memory=True) as prof:
> ```

NivekT committed:
Update on "[DataPipe] Disable profiler for IterDataPipe by default"


Previously, by default, every call to `iter` and `next` of `IterDataPipe` will have `torch.autograd.profiler.record_function()` as the context. This creates overhead and somewhat slows `IterDataPipe` down.

This PR proposes to disable that by default with a new function to enable it as necessary. I am open to the possibility that perhaps the default should `True` (enabled). Let me know what potential downside this may causes beyond what I understand.

Differential Revision: [D36830074](https://our.internmc.facebook.com/intern/diff/D36830074)

[ghstack-poisoned]

NivekT commented:
> Is that possible for TorchData to re-use the context manager of
> 
> > ```python
> > with profiler.profile(with_stack=True, profile_memory=True) as prof:
> > ```

I can make a global variable `_ITERDP_PROFILER = None` and have a function API to allow user to set it.

Then, `hook_iterator` will use that profiler if the variable `is not None`. How does that sound?

One thing on my mind is that we might be entering the context multiple times?

Once here before DataLoader is iterated:
https://github.com/pytorch/kineto/blob/plugin/vnext/tb_plugin/examples/datapipe_example.py#L24

Then again while DataLoader is iterated via `hook_iterator` of the DataPipe:
https://github.com/pytorch/kineto/blob/plugin/vnext/tb_plugin/examples/datapipe_example.py#L37

Will that be an issue?

ejguan commented:
After digging into the PyTorch codebase, we might be able to rely on `torch.autograd._profiler_enabled` to determine whether or not we should enable the profiler within `hook_iterator`.
I think within `with profiler.profile(with_stack=True, profile_memory=True) as prof:`, `torch.autograd._profiler_enabled` should return `True`

Does that make the code simpler?

NivekT committed:
Update on "[DataPipe] Disable profiler for IterDataPipe by default"


Previously, by default, every call to `iter` and `next` of `IterDataPipe` will have `torch.autograd.profiler.record_function()` as the context. This creates overhead and somewhat slows `IterDataPipe` down.

This PR proposes to disable that by default with a new function to enable it as necessary. I am open to the possibility that perhaps the default should `True` (enabled). Let me know what potential downside this may causes beyond what I understand.

Differential Revision: [D36830074](https://our.internmc.facebook.com/intern/diff/D36830074)

[ghstack-poisoned]

NivekT commented:
This PR has been updated. Now, `IterDataPipe` will only call `torch.autograd.profiler.record_function(profile_name)` if `torch.autograd._profiler_enabled() == True`. I have verified that it works with `with profiler.profile(with_stack=True, profile_memory=True) as prof:`. There is no need for users to call any data specific API to enable the profiling as long as a `torch` profiler is enabled.

The check will be performed once when the iterator is created from the DataPipe, rather than whenever `__next__` is called. Let me know if there is a need to check at every `__next__` call.

I have also decided against using `contextlib.nullcontext` for a slight performance boost.

The performance results are as follow:

```python
from torchdata.datapipes.iter import IterableWrapper
from timeit import timeit
n = 10000


def do_math(x):
    return ((x * 10 + 10) % 5 + 5) / 0.5 * 2.5 / 4.5 * 1000


datapipes = [("IterableWrapper", IterableWrapper(range(n))),
             ("Multiple MapDPs", IterableWrapper(range(n)).map(lambda x: x + 1).map(lambda x: x + 1).map(lambda x: x + 1).map(lambda x: x + 1)),
             ("MapDP(do_math)", IterableWrapper(range(n)).map(do_math))]

for name, dp in datapipes:
    res = timeit(lambda: list(dp), number=100)
    print(f"{name}:", '{0:.3g}'.format(res))
```
```
# Before any code change
# IterableWrapper: 10.6
# Multiple MapDPs: 54.3
# MapDP(do_math): 21.1

# -------------------------
# Only call `torch.autograd.profiler.record_function` if `torch.autograd._profiler_enabled()`

# Checking if `torch.autograd._profiler_enabled()` during every `__next__` and use `contextlib.nullcontext`
# IterableWrapper: 1.98
# Multiple MapDPs: 10.7
# MapDP(do_math): 4.22

# Checking if `torch.autograd._profiler_enabled()` during every `__next__` and WITHOUT `contextlib.nullcontext`
# IterableWrapper: 1.3
# Multiple MapDPs: 8.4
# MapDP(do_math): 3.32

# Checking if `torch.autograd._profiler_enabled()` ONCE during `__iter__` and WITHOUT `contextlib.nullcontext`
# IterableWrapper: 1.24
# Multiple MapDPs: 8.11
# MapDP(do_math): 3.15
```

ejguan commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -477,14 +477,19 @@ def __init__(self, iterator, source_dp, iterator_id):
             self.iterator = iterator
             self.source_dp = source_dp
             self.iterator_id = iterator_id
+            self._profiler_enabled = torch.autograd._profiler_enabled()
```
Yeah. I think there should be a special case here.
```py
it = iter(dp)  # _profiler_enabled = False

with profiler.profiler() as prof:
    for d in it:  # Profiler should be enabled but not
       ...
```

NivekT replied:
Sure, it doesn't work if you start the iteration of DataPipe outside of the profiler context:
```python
import torch
from torchdata.datapipes.iter import IterableWrapper

dp = IterableWrapper(range(5))
it = iter(dp)
print(next(it))
with torch.profiler.profile(with_stack=True, profile_memory=True) as prof:
    print(next(it))
```
It will return this:
```
0
1
```
But if you start the iteration inside:
```python
dp = IterableWrapper(range(5))
it = iter(dp)
with torch.profiler.profile(with_stack=True, profile_memory=True) as prof:
    print(next(it))
    print(next(it))
```
It returns:
```
ENABLED
0
ENABLED
1
```

In the most common use case, it should look like:
```python
dp = IterableWrapper(range(5))
with torch.profiler.profile(with_stack=True, profile_memory=True) as prof:
    it = iter(dp)
    print(next(it))
    print(next(it))
```
It returns:
```
ENABLED
0
ENABLED
1
```

NivekT replied:
@ejguan That's what I expected too, but I just tested it by modifying and got the above result. 
```python
def profiler_record_fn_context():
    print("ENABLED")
    return torch.autograd.profiler.record_function(profile_name)
```

NivekT replied:
The only reason I can think of is when this line is executed for every `__next__` but I doubt it 

https://github.com/pytorch/pytorch/blob/3415a4c663f1ed8dc2b0f3622066782b7731ac1d/torch/utils/data/datapipes/_typing.py#L508

ejguan replied:
Yeah

NivekT replied:
If we worry about users creating the iterator  outside of the profiler (not the DataPipe, creating that outside is fine) , then we can run `torch.autograd._profiler_enabled()` at every `__next__`.

ejguan commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -500,12 +505,22 @@ def wrap_generator(*args, **kwargs):
             gen = func(*args, **kwargs)
             datapipe = args[0]
             iterator_id = _set_datapipe_valid_iterator_id(datapipe)  # This ID is tied to each created iterator
+            _profiler_enabled = torch.autograd._profiler_enabled()
```
Probably because this line is only executed after the first `next`

VitalyFedyunin reviewed (approved):
I think we are safe to capture profiler state at the beginning of iterator. 
I also recommend to check how it will work in MultiProcessing.

NivekT commented:
@pytorchbot merge -g

linminhtoo commented:
Love it!
